designs/2020-timing-list-size/README.md

```diff
@@ -0,0 +1,245 @@
+- Authors: [bmish](https://github.com/bmish)
+
+# Configurable List Size For Per-Rule Performance Metrics
+
+## Summary
+
+Running `TIMING=1 eslint` from the command-line outputs a list of [per-rule performance metrics](https://eslint.org/docs/developer-guide/working-with-rules#per-rule-performance-1). The list is currently limited to the top 10 longest-running rules. We would like to enable users to see a longer list of rules if desired.
+
+Sample existing output:
+
+```sh
+$ TIMING=1 eslint lib
+Rule                    | Time (ms) | Relative
+:-----------------------|----------:|--------:
+no-multi-spaces         |    52.472 |     6.1%
+camelcase               |    48.684 |     5.7%
+no-irregular-whitespace |    43.847 |     5.1%
+valid-jsdoc             |    40.346 |     4.7%
+handle-callback-err     |    39.153 |     4.6%
+space-infix-ops         |    35.444 |     4.1%
+no-undefined            |    25.693 |     3.0%
+no-shadow               |    22.759 |     2.7%
+no-empty-class          |    21.976 |     2.6%
+semi                    |    19.359 |     2.3%
+```
+
+<!-- One-paragraph explanation of the feature. -->
+
+## Motivation
+
+Optimizing linting performance can become important in a large codebase or with particularly expensive rules. In order to determine what rules to focus on for performance optimizations, it's necessary to measure the running times of various rules while running linting on a codebase.
+
+While the current output of performance metrics for the top 10 longest-running rules helps, it doesn't give the complete picture, especially when there may be hundreds of rules enabled (from ESLint core and third-party plugins).
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+## Detailed Design
+
+We propose reusing the existing the `TIMING` environment variable to also specify the desired list size.
+
+- The performance list will continue to show only when the `TIMING` variable is present (no change).
+- When the list shows, we will read the integer value of the `TIMING` variable to determine the number of rules to show in the list.
+- A minimum list size of 10 will be used to maintain backwards compatability (and because there's not much reason to shorten the list beyond that anyway).
+- A special string value of `all` will remove the limit entirely (and make it clear that the user does not want a limit).
+
+| Command | Behavior |
+| --- | --- |
+| `eslint` | do not show |
+| `TIMING=0 eslint` | show the first 10 rules (due to minimum of 10) |
+| `TIMING=1 eslint` | show the first 10 rules (due to minimum of 10) |
+| `TIMING=5 eslint` | show the first 10 rules (due to minimum of 10) |
+| `TIMING=10 eslint` | show the first 10 rules |
+| `TIMING=11 eslint` | show the first 11 rules |
+| `TIMING=15 eslint` | show the first 15 rules |
+| `TIMING=100 eslint` | show the first 100 rules |
+| `TIMING=all eslint` | show all the rules |
```

Comment:
Should we account for other values that cannot be parsed? I've seen `TIMING=true` usage in issues, and it currently works the same as `TIMING=1`.

Reply:
Added `TIMING=true` to the list.